transport-native-epoll/src/main/c/netty_epoll_native.c

```diff
@@ -346,6 +346,59 @@ static jint netty_epoll_native_sendmmsg0(JNIEnv* env, jclass clazz, jint fd, jbo
     return (jint) res;
 }
 
+static jint netty_epoll_native_recvmmsg0(JNIEnv* env, jclass clazz, jint fd, jboolean ipv6, jobjectArray packets, jint offset, jint len) {
+    struct mmsghdr msg[len];
+    int i;
+
+    memset(msg, 0, sizeof(msg));
+
+    for (i = 0; i < len; i++) {
+        jobject packet = (*env)->GetObjectArrayElement(env, packets, i + offset);
+        msg[i].msg_hdr.msg_iov = (struct iovec*) (intptr_t) (*env)->GetLongField(env, packet, packetMemoryAddressFieldId);
+        msg[i].msg_hdr.msg_iovlen = (*env)->GetIntField(env, packet, packetCountFieldId);
+
+        struct sockaddr_storage addr;
+        msg[i].msg_hdr.msg_name = (struct sockaddr*) &addr;
+        msg[i].msg_hdr.msg_namelen = (socklen_t) sizeof(addr);
+    }
+
+    ssize_t res;
+    int err;
+    do {
+       res = recvmmsg(fd, msg, len, 0, NULL);
+       // keep on writing if it was interrupted
+    } while (res == -1 && ((err = errno) == EINTR));
+
+    if (res < 0) {
+        return -err;
+    }
+
+    for (i = 0; i < res; i++) {
+        jobject packet = (*env)->GetObjectArrayElement(env, packets, i + offset);
+        jbyteArray address = (jbyteArray) (*env)->GetObjectField(env, packet, packetAddrFieldId);
+
+        (*env)->SetIntField(env, packet, packetCountFieldId, msg[i].msg_len);
+
+        struct sockaddr_storage* addr = (struct sockaddr_storage*) msg[i].msg_hdr.msg_name;
```

Comment:
Replace with `struct sockaddr_in* ipaddr = (struct sockaddr_in*)  msg[i].msg_hdr.msg_name;`

Reply:
hmm.. but wouldn't this be kind of strange in the case of `AF_INET6` ?

Reply:
It is less readable but I don't think will make things to crash unless you access to the casted thing without any check: from a readable pov your're right anyway, so it is fine for me if you let it as you've written